lib/gateway-kernel.coffee

```diff
@@ -0,0 +1,97 @@
+child_process = require 'child_process'
+path = require 'path'
+
+_ = require 'lodash'
+uuid = require 'uuid'
+services = require('./jupyter-js-services-shim')
+
+Kernel = require './kernel'
+
+class GatewayKernel extends Kernel
+    constructor: (kernelSpec, session) ->
+        super kernelSpec, true
+
+        @session = session
+
+    interrupt: ->
+        @session.kernel.interrupt()
+
+    restart: ->
+        @session.kernel.restart()
+
+    _onIOPub: (message, onResults) ->
+        msg_type = message.header.msg_type
+
+        if msg_type is 'status'
+            status = message.content.execution_state
+            @statusView.setStatus status
+
+            # TODO(nikita): do watch callbacks correctly
+            #msg_id = message.parent_header?.msg_id
+            #if status is 'idle' and msg_id?.startsWith 'execute'
+            #    @watchCallbacks.forEach (watchCallback) ->
+            #        watchCallback()
+
+        if onResults?
+            console.log('got msg')
+            console.log(message)
+            result = @_parseIOMessage(message)
+            if result?
+                onResults result
+
+
+    execute: (code, onResults) ->
+        future = @session.kernel.execute(
+            code: code
+        )
+
+        future.onIOPub = (message) =>
+            @_onIOPub(message, onResults)
+
+        future.onReply = (message) ->
+            if message.content.status is 'error'
+                return
+            result =
+                data: 'ok'
+                type: 'text'
+                stream: 'status'
+            onResults(result)
+
+    executeWatch: (code, onResults) ->
+        # TODO(nikita): implement watches
+        # Also consider switching to user_expressions for watches
+        console.log 'Watches are not supported by remote kernels'
+
+    complete: (code, onResults) ->
+        @session.kernel.complete(
+            code: code
+            cursor_pos: code.length
+        ).then((message) ->
+            onResults message.content.matches
```

Comment:
Should be `onResults message.content` instead, otherwise autocomplete doesn't work.